compiler/rustc_arena/src/lib.rs

```diff
@@ -114,6 +117,102 @@ impl<T> Default for TypedArena<T> {
     }
 }
 
+#[inline]
+unsafe fn write_from_iter<'a, T, I: Iterator<Item = T>>(
+    mut iter: I,
+    len: usize,
+    mem: *mut T,
+) -> &'a mut [T] {
+    let mut i = 0;
+    // Use a manual loop since LLVM manages to optimize it better for
+    // slice iterators
+    loop {
+        let value = iter.next();
+        if i >= len || value.is_none() {
+            // We only return as many items as the iterator gave us, even
+            // though it was supposed to give us `len`
+            return slice::from_raw_parts_mut(mem, i);
+        }
+        ptr::write(mem.add(i), value.unwrap());
+        i += 1;
+    }
+}
+
+trait IterExt<T> {
+    fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T];
+}
+
+impl<I, T> IterExt<T> for I
+where
+    I: IntoIterator<Item = T>,
+{
+    #[inline]
+    default fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T] {
+        let mut vec: SmallVec<[_; 8]> = self.into_iter().collect();
+        if vec.is_empty() {
+            return &mut [];
+        }
+        // Move the content to the arena by copying it and then forgetting
+        // the content of the SmallVec
+        unsafe {
+            let len = vec.len();
+            let start_ptr = arena.alloc_raw_slice(len);
+            vec.as_ptr().copy_to_nonoverlapping(start_ptr, len);
+            vec.set_len(0);
+            slice::from_raw_parts_mut(start_ptr, len)
+        }
+    }
+}
+
+impl<T, const N: usize> IterExt<T> for std::array::IntoIter<T, N> {
+    #[inline]
+    fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T] {
+        if N == 0 {
+            return &mut [];
+        }
+        // Move the content to the arena by copying it
+        unsafe {
+            let len = self.len();
+            let start_ptr = arena.alloc_raw_slice(len);
+            write_from_iter(self, len, start_ptr)
```

Comment:
Hm, len() here is exact, right, so calling write_from_iter which is more general feels a bit off? We should probably be doing a direct memcpy here?

(It looks like `std::array::IntoIter` does not yet expose a as_slice function or similar, but that seems like something we can just add in this PR, unstably? cc @scottmcm who I seem to recall was doing some of the array iter work?)

Reply:
The `slice::Iter` has `.as_slice()` and is `AsRef<[T]>`, so having those on `array::IntoIter` seems like non-controversial pattern following, so that sounds good to me.  (Please take with appropriate salt considering I'm not on libs, but I would personally be willing to r+ it as unstable.)

Reply:
The slice function takes a copy though while the purpose of IntoIter is to do a move. Ideally we wouldn't do a copy but directly write into the arena, but I guess that won't work because the array construction might call the arena itself, and it might also panic.

Reply:
Made a PR: #79194